lib/Partitioner/Partitioner.cpp

```diff
@@ -139,6 +139,147 @@ static BFSLevel getBFSLevel(Function *F) {
+       ++it) {
+    GraphMemInfo cost = getGraphMemInfo((*it).second);
+    partitions.setCost((*it).first, cost);
+  }
+
+  // Move/Exchange nodes between any two connected partions, until no gain is
+  // get.
+  // Step1 Move: Assume Partition1 -> Partition2, try to move nodes from
+  // Partition2 to Partition1 if those nodes only use the nodes in
+  // Partition1(recursively) and the move won't make Partition1's memory exceeds
+  // the memory constratint, and the communication cost is minimized.
+  bool gain = true;
+  while (gain) {
+    // gain is initialized as false, it will be set to be true if there is at
+    // least one node can be moved from one set to antoher set.
+    gain = false;
+    for (PartitionNodesMap::iterator it = nodesSet.begin();
+         it != nodesSet.end(); ++it) {
+      NodesSet nSet = (*it).second;
+      std::vector<Node *> outUsers = getOutUsersWithOnePre(nSet);
+      if (outUsers.empty()) {
+        continue;
+      }
+      Function *cur = (*it).first;
+      uint64_t mem =
+          partitions.getCost(cur).constMem + partitions.getCost(cur).inMem;
+      uint64_t communicationCost = partitions.getCost(cur).outMem;
+      // Check if a node can be moved to current node set (i.e nSet).
+      for (int i = 0, e = outUsers.size(); i < e; i++) {
+        // Rule 1: this move won't break memory constraint.
+        if (memUsage_[outUsers[i]] + mem > availableMemory) {
+          continue;
+        }
+        // Rule 2: this move won't cause constant dupication.
+        bool cont = false;
+        for (int j = 0, e1 = outUsers[i]->getNumInputs(); j < e1; j++) {
+          auto in = outUsers[i]->getNthInput(j);
+          if (isa<Storage>(in.getNode()) && !in.hasOneUse()) {
+            cont = true;
+            break;
+          }
+        }
+        if (cont) {
+          continue;
+        }
+        // Rule 3: this move won't increase communication cost. Even this move
+        // won't change communication cost, according to rule 1 and rule 2, the
+        // memory comsuption of the partition where this node (i.e outUsers[i])
+        // belongs can be reduced. Therefore, it may trigger later node movement
+        // or paritionCombine.
+        nSet.insert(outUsers[i]);
+        GraphMemInfo cost = getGraphMemInfo(nSet);
+        if (cost.outMem <= communicationCost) {
+          // Move this node to current node set.
+          nSet.insert(outUsers[i]);
+          nodesSet[cur].insert(outUsers[i]);
+          Function *suc = partitions[outUsers[i]];
+          nodesSet[suc].erase(outUsers[i]);
+          // Update the partitions.
+          partitions.add(outUsers[i], cur);
+          partitions.setCost(cur, cost);
+          if (nodesSet[suc].empty()) {
+            // It is possible that after moving a node from Partition2 to
+            // Partition1, Partition2 become empty. Remove the empty partition.
+            partitions.deletePartition(suc);
```

Comment:
Here you could reuse the partition removal logic if it would be a separate method.

Reply:
here `nodesSet` is not `partitions`'s member, and this is just one circumstance under which a partition need to  be removed. Therefore, I only add `deletePartition` as a method.